Zend/zend_compile.c

```diff
@@ -2974,7 +2988,234 @@ void zend_compile_static_call(znode *result, zend_ast *ast, uint32_t type) /* {{
+		}
+
+		zend_compile_class_ref(&extends_node, extends_ast TSRMLS_CC);
+	}
+
+	opline = get_next_op(CG(active_op_array) TSRMLS_CC);
+	zend_make_var_result(&declare_node, opline TSRMLS_CC);
+
+	// TODO.AST drop this
+	GET_NODE(&CG(implementing_class), opline->result);
+
+	opline->op2_type = IS_CONST;
+	LITERAL_STR(opline->op2, lcname);
+
+	if (extends_ast) {
+		opline->opcode = ZEND_DECLARE_INHERITED_CLASS;
+		opline->extended_value = extends_node.u.op.var;
+	} else {
+		opline->opcode = ZEND_DECLARE_CLASS;
+	}
+
+	{
+		zend_string *key = zend_build_runtime_definition_key(lcname, decl->lex_pos TSRMLS_CC);
+
+		opline->op1_type = IS_CONST;
+		LITERAL_STR(opline->op1, key);
+
+		zend_hash_update_ptr(CG(class_table), key, ce);
+	}
+
+	CG(active_class_entry) = ce;
+
+	if (implements_ast) {
+		zend_compile_implements(&declare_node, implements_ast TSRMLS_CC);
+	}
+
+	zend_compile_stmt(stmt_ast TSRMLS_CC);
+
+	if (ce->constructor) {
+		ce->constructor->common.fn_flags |= ZEND_ACC_CTOR;
+		if (ce->constructor->common.fn_flags & ZEND_ACC_STATIC) {
+			zend_error_noreturn(E_COMPILE_ERROR, "Constructor %s::%s() cannot be static",
+				ce->name->val, ce->constructor->common.function_name->val);
+		}
+	}
+	if (ce->destructor) {
+		ce->destructor->common.fn_flags |= ZEND_ACC_DTOR;
+		if (ce->destructor->common.fn_flags & ZEND_ACC_STATIC) {
+			zend_error_noreturn(E_COMPILE_ERROR, "Destructor %s::%s() cannot be static",
+				ce->name->val, ce->destructor->common.function_name->val);
+		}
+	}
+	if (ce->clone) {
+		ce->clone->common.fn_flags |= ZEND_ACC_CLONE;
+		if (ce->clone->common.fn_flags & ZEND_ACC_STATIC) {
+			zend_error_noreturn(E_COMPILE_ERROR, "Clone method %s::%s() cannot be static",
+				ce->name->val, ce->clone->common.function_name->val);
+		}
+	}
+
+	/* Check for traits and proceed like with interfaces.
+	 * The only difference will be a combined handling of them in the end.
+	 * Thus, we need another opcode here. */
+	if (ce->num_traits > 0) {
+		ce->traits = NULL;
+		ce->num_traits = 0;
+		ce->ce_flags |= ZEND_ACC_IMPLEMENT_TRAITS;
+
+		zend_emit_op(NULL, ZEND_BIND_TRAITS, &declare_node, NULL TSRMLS_CC);
+	}
+
+	if (!(ce->ce_flags & (ZEND_ACC_INTERFACE|ZEND_ACC_EXPLICIT_ABSTRACT_CLASS))
+		&& (extends_ast || ce->num_interfaces > 0)
+	) {
+		zend_verify_abstract_class(ce TSRMLS_CC);
+		if (ce->num_interfaces && !(ce->ce_flags & ZEND_ACC_IMPLEMENT_TRAITS)) {
+			zend_emit_op(NULL, ZEND_VERIFY_ABSTRACT_CLASS, &declare_node, NULL TSRMLS_CC);
+		}
+	}
+
+	/* Inherit interfaces; reset number to zero, we need it for above check and
+	 * will restore it during actual implementation.
+	 * The ZEND_ACC_IMPLEMENT_INTERFACES flag disables double call to
+	 * zend_verify_abstract_class() */
+	if (ce->num_interfaces > 0) {
+		ce->interfaces = NULL;
+		ce->num_interfaces = 0;
+		ce->ce_flags |= ZEND_ACC_IMPLEMENT_INTERFACES;
+	}
+
+	CG(active_class_entry) = active;
+	
+	return ce;
+}
+/* }}} */
```

Comment:
I'd suggest keeping this code where it was and add a forward declaration to it - it's pretty hard to see what changed otherwise ^^